File: backend/services/stockRequestService.js
// backend/services/stockRequestService.js
import db from "../config/db.js";
import { processBatchMovementsService, processBatchOpnameService } from "./stockService.js";
import { emitSharedTaskSignal } from "./firebaseSignalService.js";
import * as stockRequestRepo from "../repositories/stockRequestRepository.js";
import * as notificationService from "./notificationService.js";

/**
 * Membuat permintaan stok (bulk)
 */
export const createStockRequestService = async (payload, userId) => {
  const connection = await db.getConnection();
  await connection.beginTransaction();
  try {
    const fullPayload = { ...payload, requesterId: userId };
    const type = payload.type || "TRANSFER";

    // Validasi basic
    if (type === "TRANSFER") {
      if (!payload.fromLocationId || !payload.toLocationId) {
        throw new Error("Lokasi asal dan tujuan harus diisi untuk transfer.");
      }
      if (payload.fromLocationId === payload.toLocationId) {
        throw new Error("Lokasi asal dan tujuan tidak boleh sama.");
      }
    } else if (type === "STOCK_OPNAME") {
      if (!payload.toLocationId) {
        throw new Error("Lokasi opname harus diisi.");
      }
    }

    if (!payload.items || payload.items.length === 0) {
      throw new Error("Minimal harus ada satu produk yang diminta.");
    }

    const requestId = await stockRequestRepo.createStockRequest(connection, fullPayload);
    await connection.commit();
    emitSharedTaskSignal("STOCK_REQUESTS", "REFRESH_REQUESTS").catch((e) => console.error(e));

    // Notifikasi ke user dengan permission approve-stock-requests
    notificationService
      .notifyUsersByPermission(
        "approve-stock-requests",
        "WMS",
        "Permintaan Stok Baru",
        `Permintaan stok baru (${type}) telah dibuat dan menunggu persetujuan.`,
        { requestId, type },
        userId,
      )
      .catch((e) => console.error("[NOTIFY_ERROR]", e));

    return { success: true, message: "Permintaan stok berhasil dibuat.", requestId };
  } catch (error) {
    await connection.rollback();
    throw error;
  } finally {
    connection.release();
  }
};

/**
 * Mendapatkan semua request dengan filter
 */
export const getStockRequestsService = async (filters) => {
  const connection = await db.getConnection();
  try {
    const result = await stockRequestRepo.getStockRequests(connection, filters);
    return {
      success: true,
      data: result.data,
      pagination: result.pagination,
    };
  } finally {
    connection.release();
  }
};

/**
 * Approve Permintaan Stok (Oleh Pihak Pengirim)
 */
export const approveStockRequestService = async (requestId, user) => {
  const connection = await db.getConnection();
  await connection.beginTransaction();
  try {
    const request = await stockRequestRepo.getStockRequestById(connection, requestId);
    if (!request) throw new Error("Permintaan stok tidak ditemukan.");
    if (request.status !== "PENDING")
      throw new Error("Hanya permintaan berstatus PENDING yang dapat disetujui.");

    // Cegah self-approval
    if (request.requester_id === user.id) {
      throw new Error("Anda tidak dapat menyetujui permintaan Anda sendiri.");
    }

    if (request.type === "STOCK_OPNAME") {
      const movements = request.items.map((reqItem) => ({
        sku: reqItem.sku,
        quantity: reqItem.quantity,
        toLocationId: request.to_location_id,
        notes: `Stock Request Opname ${request.request_number}`,
      }));

      await processBatchOpnameService({
        movements: movements,
        userId: user.id,
        userRoleId: user.role_id || 1,
      });

      for (const reqItem of request.items) {
        await stockRequestRepo.updateRequestItemReceived(connection, reqItem.id, reqItem.quantity);
      }

      await stockRequestRepo.updateStockRequestStatus(connection, requestId, "COMPLETED");
      await connection.commit();
      emitSharedTaskSignal("STOCK_REQUESTS", "REFRESH_REQUESTS").catch((e) => console.error(e));

      // Notifikasi ke requester
      notificationService
        .notifyUsers(
          [request.requester_id],
          "WMS",
          "Permintaan Opname Disetujui",
          `Permintaan stok opname ${request.request_number} telah disetujui dan stok disesuaikan.`,
          { requestId, requestNumber: request.request_number },
          true,
        )
        .catch((e) => console.error("[NOTIFY_ERROR]", e));

      return { success: true, message: "Permintaan stok opname disetujui dan stok disesuaikan." };
    }

    await stockRequestRepo.updateStockRequestStatus(connection, requestId, "APPROVED");

    await connection.commit();
    emitSharedTaskSignal("STOCK_REQUESTS", "REFRESH_REQUESTS").catch((e) => console.error(e));

    // Notifikasi ke requester
    notificationService
      .notifyUsers(
        [request.requester_id],
        "WMS",
        "Permintaan Stok Disetujui",
        `Permintaan stok ${request.request_number} telah disetujui. Silakan lakukan konfirmasi penerimaan.`,
        { requestId, requestNumber: request.request_number },
        true,
      )
      .catch((e) => console.error("[NOTIFY_ERROR]", e));

    return { success: true, message: "Permintaan stok berhasil disetujui." };
  } catch (error) {
    await connection.rollback();
    throw error;
  } finally {
    connection.release();
  }
};

/**
 * Reject Permintaan Stok
 */
export const rejectStockRequestService = async (requestId, _) => {
  const connection = await db.getConnection();
  await connection.beginTransaction();
  try {
    const request = await stockRequestRepo.getStockRequestById(connection, requestId);
    if (!request) throw new Error("Permintaan stok tidak ditemukan.");
    if (request.status !== "PENDING")
      throw new Error("Hanya permintaan berstatus PENDING yang dapat ditolak.");

    await stockRequestRepo.updateStockRequestStatus(connection, requestId, "REJECTED");

    await connection.commit();
    emitSharedTaskSignal("STOCK_REQUESTS", "REFRESH_REQUESTS").catch((e) => console.error(e));

    // Notifikasi ke requester
    notificationService
      .notifyUsers(
        [request.requester_id],
        "WMS",
        "Permintaan Stok Ditolak",
        `Permintaan stok ${request.request_number} telah ditolak.`,
        { requestId, requestNumber: request.request_number },
        true,
      )
      .catch((e) => console.error("[NOTIFY_ERROR]", e));

    return { success: true, message: "Permintaan stok telah ditolak." };
  } catch (error) {
    await connection.rollback();
    throw error;
  } finally {
    connection.release();
  }
};

/**
 * Complete Permintaan Stok (Oleh Pihak Penerima)
 * Ini akan mentrigger transfer stok.
 */
export const completeStockRequestService = async (requestId, receivedItems, user) => {
  const connection = await db.getConnection();
  await connection.beginTransaction();
  try {
    const request = await stockRequestRepo.getStockRequestById(connection, requestId);
    if (!request) throw new Error("Permintaan stok tidak ditemukan.");
    if (request.status !== "APPROVED")
      throw new Error("Hanya permintaan berstatus APPROVED yang dapat diselesaikan.");

    // Update received_quantity
    const movements = [];
    for (const reqItem of request.items) {
      // Cari received quantity dari payload receivedItems (jika ada), kalau tidak full quantity
      const match = receivedItems
        ? receivedItems.find((r) => r.productId === reqItem.product_id)
        : null;
      const rQty = match ? match.receivedQuantity : reqItem.quantity;

      await stockRequestRepo.updateRequestItemReceived(connection, reqItem.id, rQty);

      if (request.type === "STOCK_OPNAME") {
        throw new Error("Permintaan Stock Opname tidak memerlukan konfirmasi penerimaan.");
      } else {
        if (rQty > 0) {
          movements.push({
            sku: reqItem.sku,
            quantity: rQty,
            fromLocationId: request.from_location_id,
            toLocationId: request.to_location_id,
            notes: `Stock Request ${request.request_number}`,
          });
        }
      }
    }

    // Trigger pergerakan stok menggunakan existing service.
    if (movements.length > 0) {
      await processBatchMovementsService({
        type: "TRANSFER",
        fromLocationId: request.from_location_id,
        toLocationId: request.to_location_id,
        notes: `Penyelesaian Permintaan Stok ${request.request_number}`,
        movements: movements,
        userId: user.id,
        userRoleId: 1, // Bypass pengecekan lokasi karena ini sudah di-approve
      });
    }

    await stockRequestRepo.updateStockRequestStatus(connection, requestId, "COMPLETED");

    await connection.commit();
    emitSharedTaskSignal("STOCK_REQUESTS", "REFRESH_REQUESTS").catch((e) => console.error(e));

    // Notifikasi ke requester
    notificationService
      .notifyUsers(
        [request.requester_id],
        "WMS",
        "Permintaan Stok Selesai",
        `Permintaan stok ${request.request_number} telah selesai dan stok berhasil ditransfer.`,
        { requestId, requestNumber: request.request_number },
        true,
      )
      .catch((e) => console.error("[NOTIFY_ERROR]", e));

    return { success: true, message: "Permintaan stok selesai dan stok telah ditransfer." };
  } catch (error) {
    await connection.rollback();
    throw error;
  } finally {
    connection.release();
  }
};

/**
 * Bulk Action untuk Permintaan Stok
 * Mendukung aksi APPROVE dan REJECT secara massal.
 */
export const bulkActionStockRequestService = async (requestIds, action, user) => {
  if (!Array.isArray(requestIds) || requestIds.length === 0) {
    throw new Error("Tidak ada ID permintaan yang dipilih.");
  }

  const results = {
    successCount: 0,
    failedCount: 0,
    details: [],
  };

  for (const id of requestIds) {
    try {
      if (action === "APPROVE") {
        await approveStockRequestService(id, user);
      } else if (action === "REJECT") {
        await rejectStockRequestService(id, user);
      } else {
        throw new Error(`Aksi ${action} tidak didukung secara massal.`);
      }

      results.successCount++;
      results.details.push({ id, status: "success" });
    } catch (err) {
      results.failedCount++;
      results.details.push({ id, status: "failed", reason: err.message });
    }
  }

  return {
    success: true,
    message: `Memproses ${requestIds.length} permintaan. Berhasil: ${results.successCount}, Gagal: ${results.failedCount}.`,
    data: results,
  };
};

File: backend/internal/service/stock_request_service.go
package service

import (
	"context"
	"errors"
	"fmt"

	"github.com/dps-wmhris/backend/internal/database"
	"github.com/dps-wmhris/backend/internal/dto"
	"github.com/dps-wmhris/backend/internal/model"
	"github.com/dps-wmhris/backend/internal/repository"
	"github.com/jmoiron/sqlx"
)

type StockRequestService interface {
	CreateStockRequest(ctx context.Context, userID int, req dto.CreateStockRequest) (*model.StockRequest, error)
	GetAllStockRequests(ctx context.Context) ([]model.StockRequest, error)
	ApproveStockRequest(ctx context.Context, id int, userID int, roleID int) error
	RejectStockRequest(ctx context.Context, id int, userID int, roleID int) error
	DispatchStockRequest(ctx context.Context, id int, userID int, roleID int) error
	CompleteStockRequest(ctx context.Context, id int, req dto.CompleteStockRequest, userID int, roleID int) error
	BulkActionStockRequest(ctx context.Context, req dto.BulkActionStockRequest, userID int, roleID int) (map[string]interface{}, error)
}

type stockRequestServiceImpl struct {
	db                  *sqlx.DB
	stockRequestRepo    repository.StockRequestRepository
	stockService        StockService
	notificationService NotificationService
}

func NewStockRequestService(
	db *sqlx.DB,
	stockRequestRepo repository.StockRequestRepository,
	stockService StockService,
	notificationService NotificationService,
) StockRequestService {
	return &stockRequestServiceImpl{
		db:                  db,
		stockRequestRepo:    stockRequestRepo,
		stockService:        stockService,
		notificationService: notificationService,
	}
}

func (s *stockRequestServiceImpl) CreateStockRequest(ctx context.Context, userID int, req dto.CreateStockRequest) (*model.StockRequest, error) {
	if req.Type == "TRANSFER" {
		if req.FromLocationID == nil || req.ToLocationID == nil {
			return nil, errors.New("Lokasi asal dan tujuan harus diisi untuk transfer.")
		}
		if *req.FromLocationID == *req.ToLocationID {
			return nil, errors.New("Lokasi asal dan tujuan tidak boleh sama.")
		}
	} else if req.Type == "STOCK_OPNAME" {
		if req.ToLocationID == nil {
			return nil, errors.New("Lokasi opname harus diisi.")
		}
	}

	if len(req.Items) == 0 {
		return nil, errors.New("Minimal harus ada satu produk yang diminta.")
	}

	request := &model.StockRequest{
		RequestNumber:  req.RequestNumber,
		Type:           req.Type,
		RequesterID:    userID,
		FromLocationID: req.FromLocationID,
		ToLocationID:   req.ToLocationID,
		Status:         "PENDING",
		Notes:          req.Notes,
	}

	err := database.WithTransaction(s.db, ctx, func(tx *sqlx.Tx) error {
		if err := s.stockRequestRepo.CreateTx(ctx, tx, request); err != nil {
			return err
		}

		for _, item := range req.Items {
			reqItem := &model.StockRequestItem{
				StockRequestID:   request.ID,
				ProductID:        item.ProductID,
				Quantity:         item.Quantity,
				ReceivedQuantity: 0,
			}
			if err := s.stockRequestRepo.CreateItemTx(ctx, tx, reqItem); err != nil {
				return err
			}
			request.Items = append(request.Items, *reqItem)
		}

		return nil
	})

	if err != nil {
		return nil, err
	}

	// Notifikasi
	s.notificationService.NotifyUsersByPermission(ctx, "approve-stock-requests", "WMS", "Permintaan Stok Baru",
		fmt.Sprintf("Permintaan stok baru (%s) telah dibuat dan menunggu persetujuan.", request.Type),
		map[string]interface{}{"requestId": request.ID, "type": request.Type}, &userID, true)

	return request, nil
}

func (s *stockRequestServiceImpl) GetAllStockRequests(ctx context.Context) ([]model.StockRequest, error) {
	return s.stockRequestRepo.FindAll(ctx)
}

func (s *stockRequestServiceImpl) ApproveStockRequest(ctx context.Context, id int, userID int, roleID int) error {
	request, err := s.stockRequestRepo.FindByID(ctx, id)
	if err != nil {
		return errors.New("Permintaan stok tidak ditemukan.")
	}

	if request.Status != "PENDING" {
		return errors.New("Hanya permintaan berstatus PENDING yang dapat disetujui.")
	}

	if request.RequesterID == userID {
		return errors.New("Anda tidak dapat menyetujui permintaan Anda sendiri.")
	}

	return database.WithTransaction(s.db, ctx, func(tx *sqlx.Tx) error {
		if request.Type == "STOCK_OPNAME" {
			var movements []dto.BatchMovementRequest
			for _, item := range request.Items {
				notes := fmt.Sprintf("Stock Request Opname %s", request.RequestNumber)
				movements = append(movements, dto.BatchMovementRequest{
					SKU:          item.SKU,
					Quantity:     item.Quantity,
					ToLocationID: request.ToLocationID,
					Notes:        &notes,
				})
			}

			err = s.stockService.ProcessBatchMovements(ctx, dto.BatchProcessRequest{
				Type:           "ADJUSTMENT", 
				ToLocationID:   request.ToLocationID,
				Notes:          &request.RequestNumber,
				Movements:      movements,
			}, userID, roleID)
			
			if err != nil {
				return err
			}

			for _, item := range request.Items {
				if err := s.stockRequestRepo.UpdateItemReceivedQtyTx(ctx, tx, item.ID, item.Quantity); err != nil {
					return err
				}
			}

			if err := s.stockRequestRepo.UpdateStatusTx(ctx, tx, request.ID, "COMPLETED"); err != nil {
				return err
			}
		} else {
			if err := s.stockRequestRepo.UpdateStatusTx(ctx, tx, request.ID, "APPROVED"); err != nil {
				return err
			}
		}
		return nil
	})
}

func (s *stockRequestServiceImpl) RejectStockRequest(ctx context.Context, id int, userID int, roleID int) error {
	request, err := s.stockRequestRepo.FindByID(ctx, id)
	if err != nil {
		return errors.New("Permintaan stok tidak ditemukan.")
	}

	if request.Status != "PENDING" {
		return errors.New("Hanya permintaan berstatus PENDING yang dapat ditolak.")
	}

	return database.WithTransaction(s.db, ctx, func(tx *sqlx.Tx) error {
		return s.stockRequestRepo.UpdateStatusTx(ctx, tx, request.ID, "REJECTED")
	})
}

func (s *stockRequestServiceImpl) DispatchStockRequest(ctx context.Context, id int, userID int, roleID int) error {
	request, err := s.stockRequestRepo.FindByID(ctx, id)
	if err != nil {
		return errors.New("Permintaan stok tidak ditemukan.")
	}

	if request.Status != "APPROVED" {
		return errors.New("Hanya permintaan berstatus APPROVED yang dapat dikirim.")
	}
	if request.Type == "STOCK_OPNAME" {
		return errors.New("Permintaan Stock Opname tidak memerlukan pengiriman.")
	}

	return database.WithTransaction(s.db, ctx, func(tx *sqlx.Tx) error {
		var movements []dto.BatchMovementRequest
		for _, item := range request.Items {
			notes := fmt.Sprintf("Pengiriman Permintaan Stok %s", request.RequestNumber)
			movements = append(movements, dto.BatchMovementRequest{
				SKU:            item.SKU,
				Quantity:       item.Quantity,
				FromLocationID: request.FromLocationID,
				Notes:          &notes,
			})
		}

		err = s.stockService.ProcessBatchMovements(ctx, dto.BatchProcessRequest{
			Type:           "TRANSFER_OUT",
			FromLocationID: request.FromLocationID,
			Notes:          &request.RequestNumber,
			Movements:      movements,
		}, userID, 1) 

		if err != nil {
			return err
		}

		return s.stockRequestRepo.UpdateStatusTx(ctx, tx, request.ID, "SHIPPED")
	})
}

func (s *stockRequestServiceImpl) CompleteStockRequest(ctx context.Context, id int, req dto.CompleteStockRequest, userID int, roleID int) error {
	request, err := s.stockRequestRepo.FindByID(ctx, id)
	if err != nil {
		return errors.New("Permintaan stok tidak ditemukan.")
	}

	if request.Status != "SHIPPED" {
		return errors.New("Hanya permintaan berstatus SHIPPED yang dapat diselesaikan.")
	}

	return database.WithTransaction(s.db, ctx, func(tx *sqlx.Tx) error {
		var movements []dto.BatchMovementRequest
		for _, item := range request.Items {
			rQty := item.Quantity
			for _, rItem := range req.ReceivedItems {
				if rItem.ProductID == item.ProductID {
					rQty = rItem.ReceivedQuantity
					break
				}
			}

			if err := s.stockRequestRepo.UpdateItemReceivedQtyTx(ctx, tx, item.ID, rQty); err != nil {
				return err
			}

			if rQty > 0 {
				notes := fmt.Sprintf("Penerimaan Permintaan Stok %s", request.RequestNumber)
				movements = append(movements, dto.BatchMovementRequest{
					SKU:          item.SKU,
					Quantity:     rQty,
					ToLocationID: request.ToLocationID,
					Notes:        &notes,
				})
			}
		}

		if len(movements) > 0 {
			err = s.stockService.ProcessBatchMovements(ctx, dto.BatchProcessRequest{
				Type:         "TRANSFER_IN",
				ToLocationID: request.ToLocationID,
				Notes:        &request.RequestNumber,
				Movements:    movements,
			}, userID, 1) 

			if err != nil {
				return err
			}
		}

		return s.stockRequestRepo.UpdateStatusTx(ctx, tx, request.ID, "COMPLETED")
	})
}

func (s *stockRequestServiceImpl) BulkActionStockRequest(ctx context.Context, req dto.BulkActionStockRequest, userID int, roleID int) (map[string]interface{}, error) {
	successCount := 0
	failedCount := 0
	details := []map[string]interface{}{}

	for _, id := range req.RequestIds {
		var err error
		if req.Action == "APPROVE" {
			err = s.ApproveStockRequest(ctx, id, userID, roleID)
		} else if req.Action == "REJECT" {
			err = s.RejectStockRequest(ctx, id, userID, roleID)
		}

		if err == nil {
			successCount++
			details = append(details, map[string]interface{}{"id": id, "status": "success"})
		} else {
			failedCount++
			details = append(details, map[string]interface{}{"id": id, "status": "failed", "reason": err.Error()})
		}
	}

	return map[string]interface{}{
		"successCount": successCount,
		"failedCount":  failedCount,
		"details":      details,
	}, nil
}
